feat: fear condition

bosmak opened the pull request:
# Description

Implement fear condition.

## Type of change

  - [X] New feature (non-breaking change which adds functionality)

## Checklist

  - [X] My code follows the style guidelines of this project
  - [X] I have performed a self-review of my own code
  - [X] I checked the PR checks reports
  - [X] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [x] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works

bosmak committed:
implement fear condition

dudantas commented on src/creatures/monsters/monsters.cpp:
```diff
@@ -401,6 +401,15 @@ bool Monsters::deserializeSpell(const pugi::xml_node& node, spellBlock_t& sb, co
 
 			Condition* condition = Condition::createCondition(CONDITIONID_COMBAT, CONDITION_DRUNK, duration, 0);
 			combat->addCondition(condition);
+		} else if (tmpName == "fear") {
```
This not is more used, is from old XML monsters loading.

bosmak replied:
Should I remove this bit or the whole function?

dudantas replied:
> Should I remove this bit or the whole function?

You can remove everything if you want.

bosmak replied:
I think I will leave it as is, I tried to remove only this method but seems that there are more parts of the code that may need to change.

I believe this may be its own PR to remove this function and other related ones.

dudantas commented:
Your pull request looks great, I approved the execution of the checks. I'll take a closer look later, but from what I've seen, everything seems ok.

guispiller reviewed (commented):
Thank you, excellent PR!
Tested here, feared condition working fine restricting for party members, but still have some issues/restrictions for the condition itself, i'll point here:
- Fear condition when activated, should remain for 6 seconds;
- After being feared, when the condition finishes, you should have 10 seconds "cooldown" until you can get feared again;
- When feared, you should not be able to use any spell, nor equip amulets/rings (at the pr sometimes you can use spell, and can equip items anytime)

dudantas commented:
> Thank you, excellent PR! Tested here, feared condition working fine restricting for party members, but still have some issues/restrictions for the condition itself, i'll point here:
> 
> * Fear condition when activated, should remain for 6 seconds;
> * After being feared, when the condition finishes, you should have 10 seconds "cooldown" until you can get feared again;
> * When feared, you should not be able to use any spell, nor equip amulets/rings (at the pr sometimes you can use spell, and can equip items anytime)

I'll add the pull as a draft until the author decides to analyse it and, if necessary, fix the problems.

dudantas committed:
Merge branch 'main' into fear_condition

bosmak commented:
Sorry for the delay, I have been stuck on some work-related topics.

I will try to work on it this week!

bosmak committed:
Merge branch 'main' into fear_condition

bosmak committed:
chore: change log level on fear checks

bosmak committed:
fix: fear condition duration to 6 seconds

bosmak committed:
feat: feared players can't change equipaments

bosmak committed:
feat: add a cooldown after a player is feared

bosmak commented:
@dudantas
I tried to remove the XML function, but it seems more code is attached, so I left it as is.

@guispiller
I couldn't reproduce using a spell under the fear condition.
 Besides that, I adjusted the default duration to 6 seconds, Fixed that it was possible to change the equipment using the hotkeys, and finally added the cooldown for 10 seconds after being feared.

Could you test again?

----

For other contributors looking to test, just add the following attack to a monster

```lua
{name ="fear", interval = 2000, chance = 100, range=7, effect=CONST_ME_MAGIC_RED, target=true}
```

bosmak committed:
chore: add a second check if feared when casting a spell on playerSpellCheck function. Akk recommended this function to ensure no spell can be casted.

bosmak committed:
fix: some sonar alerts

tutbarao commented:
Tested! Functional!

jlcvp commented on src/creatures/combat/combat.cpp:
```diff
@@ -633,12 +635,59 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
+				 */
+				Party* party = player->getParty();
+				if(party) {
+					uint8_t affectedCount = std::ceil((party->getMemberCount()+5)/5);
```
To obey the condition described, it should be `std::floor` instead

e.g.:
if party member count is 3, we get `(3+5)/5`, `1.6`, which is greater than 1, thus, `std::ceil` will make it 2

jlcvp commented on src/creatures/combat/condition.cpp:
```diff
@@ -1460,6 +1496,276 @@ void ConditionDamage::generateDamageList(int32_t amount, int32_t start, std::lis
+		}
+	}
+
+	return false;
```
The returned values here are swapped, it should return `false` if any direction can be walked to, and `true` otherwise

jlcvp commented on src/creatures/combat/condition.cpp:
```diff
@@ -1460,6 +1496,276 @@ void ConditionDamage::generateDamageList(int32_t amount, int32_t start, std::lis
+				fleeIndx = 0;
+			}
+
+			if(!isStuck(creature, pos)){ // Check if it is possible to walk to any direction
```
after fixing the `isStuck` logic, this if condition must not be negated
```suggestion
			if(isStuck(creature, pos)){ // Check if it is possible to walk to any direction
```

jlcvp commented on src/creatures/combat/condition.cpp:
```diff
@@ -1460,6 +1496,276 @@ void ConditionDamage::generateDamageList(int32_t amount, int32_t start, std::lis
+		}
+		
+		return true;
+	}
```
We cannot use hardcoded indexes here, as `getRandomDirection()` function shuffles the direction array

jlcvp commented on src/creatures/combat/condition.cpp:
```diff
@@ -1460,6 +1496,276 @@ void ConditionDamage::generateDamageList(int32_t amount, int32_t start, std::lis
+}
+
+bool ConditionFeared::getRandomDirection(Creature* creature, Position pos) {
+	std::shuffle(dirList.begin(), dirList.end(), getRandomGenerator());
```
shuffling the shared vector of the object will impact in other functions that uses `fleeIndx`

jlcvp replied:
Maybe using the same strategy as in ```Monster::getRandomStep```  may solve this

jlcvp commented on src/creatures/players/player.h:
```diff
@@ -1789,6 +1789,18 @@ class Player final : public Creature, public Cylinder
+				return true;
+			}
+			return false;
+		}
```
As the `contitiontype` here will always be `CONDITION_FEARED`, we can omit that parameter
```suggestion
		void setImmuneFear() {
			fearCondition.first = CONDITION_FEARED;
			fearCondition.second = OTSYS_TIME() + 10000;
		}
		bool isImmuneFear() {
			uint64_t timenow = OTSYS_TIME();
			if ((fearCondition.first == CONDITION_FEARED)
                    && (timenow <= fearCondition.second)) {
				return true;
			}
			return false;
		}
```

jlcvp commented on src/map/map.cpp:
```diff
@@ -1140,6 +1140,13 @@ int_fast32_t AStarNodes::getTileWalkCost(const Creature& creature, const Tile* t
+		const Player* player = creature.getPlayer();
+		if (player && field && !field->isBlocking() && field->getDamage() != 0) {
+			cost += MAP_NORMALWALKCOST * 18;
+		}
```
nice addition!

jlcvp reviewed (commented):
Looking great, just a few suggestions

bosmak committed:
fix: wrong return on ConditionFear::isStuck

bosmak committed:
fix: omit parameters on setImmuneFear and isImmuneFear functions

bosmak committed:
fix: fear being applied to more players than it should when player is in a party

bosmak committed:
fix: stop randomizing dirList on ConditionFeared

bosmak committed:
fix: codacy issues

bosmak committed:
fix: try to fix some sonar alerts

bosmak committed:
chore: improve format

jlcvp reviewed (dismissed):
Looking great now, just this annoying sonar error wrongly accusing the random generator

beats-dh committed:
Merge branch 'main' into pr/542

dudantas commented:
> Looking great now, just this annoying sonar error wrongly accusing the random generator

This is why the branch is out of date.

dudantas committed:
Merge branch 'main' into fear_condition

dudantas committed:
Remove

dudantas committed:
Fix

dudantas commented on src/creatures/monsters/monsters.cpp:
```diff
@@ -399,6 +399,15 @@ bool Monsters::deserializeSpell(const pugi::xml_node& node, spellBlock_t& sb, co
+			}
+
+			Condition* condition = Condition::createCondition(CONDITIONID_COMBAT, CONDITION_FEARED, duration, 0);
+			combat->addCondition(condition);
```
```suggestion
```

dudantas committed:
Update lua_enums.hpp

dudantas committed:
Fix

dudantas committed:
Update src/creatures/monsters/monsters.cpp

bosmak commented:
@dudantas, I saw that you removed all enum declarations from the lua_enums.hpp not sure if this was expected.

Anyway, you need to add `;` on `src/lua/functions/core/game/config_functions.cpp`. I changed the macro to need it to correctly format when we implement the clang-formatter, if not the file will look like this

https://github.com/bosmak/canary/blob/35679dd98c70f814718be8628dace9aaac9722b5/src/lua/functions/core/game/config_functions.hpp

dudantas commented:
> @dudantas, I saw that you removed all enum declarations from the lua_enums.hpp not sure if this was expected.
> 
> Anyway, you need to add `;` on `src/lua/functions/core/game/config_functions.cpp`. I changed the macro to need it to correctly format when we implement the clang-formatter, if not the file will look like this
> 
> https://github.com/bosmak/canary/blob/35679dd98c70f814718be8628dace9aaac9722b5/src/lua/functions/core/game/config_functions.hpp

Hi, could you change it please? I modified it, as I was resolving the conflicts that occurred because of the merge of other prs who worked on the same thing...

dudantas committed:
Merge branch 'main' into fear_condition

dudantas commented on src/creatures/combat/combat.cpp:
```diff
@@ -611,9 +611,11 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
+		
 		Player* player = target->getPlayer();
 		if (player) {
+			//Cleanse charm rune (target as player)
```
```suggestion
			// Cleanse charm rune (target as player)
```

dudantas commented on src/creatures/combat/combat.cpp:
```diff
@@ -634,12 +636,59 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
+			/**
+			 * Specifics checks for Feared condtion
+			 */
+			if(condition->getType() == CONDITION_FEARED){
```
```suggestion
			if (condition->getType() == CONDITION_FEARED) {
```

dudantas commented on src/creatures/combat/combat.cpp:
```diff
@@ -634,12 +636,59 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
+				 *
+				 */
+				Party* party = player->getParty();
+				if(party) {
```
```suggestion
				if (party) {
```

dudantas commented on src/creatures/combat/combat.cpp:
```diff
@@ -634,12 +636,59 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
+						}
+					}
+
+					if(affectedCount <= 0){
```
```suggestion
					if (affectedCount <= 0) {
```

dudantas commented on src/creatures/combat/combat.cpp:
```diff
@@ -634,12 +636,59 @@ void Combat::CombatConditionFunc(Creature* caster, Creature* target, const Comba
+					int8_t affectedCount = std::floor((party->getMemberCount()+5)/5);
+					SPDLOG_DEBUG("[Combat::CombatConditionFunc] Player is member of a party, {} members can be feared", affectedCount);
+					
+					for (Player* member : party->getMembers()){
```
```suggestion
					for (Player* member : party->getMembers()) {
```

bosmak committed:
chore: fix formating for config_functions and lua_enums

bosmak commented:
@dudantas Done.

I needed to comment the forge related bits from `config_functions.cpp` to be able to build, but committed them as is, not sure why you added them before they are on main, but I left as is.

```
	registerEnumIn(L, "configKeys", FORGE_COST_ONE_SLIVER);
	registerEnumIn(L, "configKeys", FORGE_SLIVER_AMOUNT);
	registerEnumIn(L, "configKeys", FORGE_CORE_COST);
	registerEnumIn(L, "configKeys", FORGE_MAX_DUST);
	registerEnumIn(L, "configKeys", FORGE_FUSION_DUST_COST);
	registerEnumIn(L, "configKeys", FORGE_TRANSFER_DUST_COST);
	registerEnumIn(L, "configKeys", FORGE_BASE_SUCCESS_RATE);
	registerEnumIn(L, "configKeys", FORGE_BONUS_SUCCESS_RATE);
	registerEnumIn(L, "configKeys", FORGE_TIER_LOSS_REDUCTION);
```

dudantas commented:
> I needed to comment the forge related bits from `config_functions.cpp` to be able to build, but committed them as is, not sure why you added them before they are on main, but I left as is.

Oops, it was a small mistake of mine, because I'm currently in the forge branch and I went to compare with what was missing, I'll remove it, thanks.

dudantas committed:
Remove not needed

dudantas committed:
Merge branch 'main' into fear_condition

bosmak committed:
Merge branch 'main' into fear_condition

Olimpotibia commented:
Fear is working at the moment?

Olimpotibia commented:
Anyone can open new branch to test with the last main version? (*with sound changes*)

thanks you

murilo09 commented:
> Anyone can open new branch to test with the last main version? (_with sound changes_)
> 
> thanks you


#531 is not the last main version. Both are PRs and should be tested separately.

bosmak committed:
Merge remote-tracking branch 'origin/main' into fear_condition

github-actions[bot] committed:
Code format - (Clang-format)

bosmak committed:
fix: issue on build

dudantas committed:
Merge branch 'main' into fear_condition

dudantas commented:
@bosmak we are merging the pulls by priority, at the moment I'm on the protocol 13x pull, soon we'll get to yours, I just ask for a little more patience, ok? Thanks

paczielny commented:
Severity	Code	Description	Project	File	Line	Suppression State
Error	C3668	'ConditionFeared::executeCondition': method with override specifier 'override' did not override any base class methods	

when try build :/

dudantas commented:
> Severity Code Description Project File Line Suppression State Error C3668 'ConditionFeared::executeCondition': method with override specifier 'override' did not override any base class methods
> 
> when try build :/

you must be using some mixture of prs or something, because the compilation checks are passing normally

paczielny commented:
> > Severity Code Description Project File Line Suppression State Error C3668 'ConditionFeared::executeCondition': method with override specifier 'override' did not override any base class methods
> > when try build :/
> 
> you must be using some mixture of prs or something, because the compilation checks are passing normally

maybe version? 13.11

mohuga commented:
character dont walk after feared, just stuck! but conditions of heal, time, etc still working correctly

dudantas commented:
> > > Severity Code Description Project File Line Suppression State Error C3668 'ConditionFeared::executeCondition': method with override specifier 'override' did not override any base class methods
> > > when try build :/
> > 
> > 
> > you must be using some mixture of prs or something, because the compilation checks are passing normally
> 
> maybe version? 13.11

You need to test on the pull request branch only. The error you have is not related to the pull request itself, but its implementation.

elsongabriel commented:
> > > > Severity Code Description Project File Line Suppression State Error C3668 'ConditionFeared::executeCondition': method with override specifier 'override' did not override any base class methods
> > > > when try build :/
> > > 
> > > 
> > > you must be using some mixture of prs or something, because the compilation checks are passing normally
> > 
> > 
> > maybe version? 13.11
> 
> You need to test on the pull request branch only. The error you have is not related to the pull request itself, but its implementation.

I merged 3 branches.. this (fear_condition), test (client 13.11), and missing_bosses, it works!
we have to pay attention when merging with PR client 13, to use the functions with int64_t parameters.
@paczielny

dudantas committed:
Merge branch 'main' into fear_condition

dudantas commented:
> > > > > Severity Code Description Project File Line Suppression State Error C3668 'ConditionFeared::executeCondition': method with override specifier 'override' did not override any base class methods
> > > > > when try build :/
> > > > 
> > > > 
> > > > you must be using some mixture of prs or something, because the compilation checks are passing normally
> > > 
> > > 
> > > maybe version? 13.11
> > 
> > 
> > You need to test on the pull request branch only. The error you have is not related to the pull request itself, but its implementation.
> 
> I merged 3 branches.. this (fear_condition), test (client 13.11), and missing_bosses, it works! we have to pay attention when merging with PR client 13, to use the functions with int64_t parameters. @paczielny

Do you consider that the system is ok to be merged? Now we no longer have the changes to int64_t, which should make it easier for the system to implement.

paczielny commented:
how add fear condition in attack of monsters?

paczielny commented:
> how add fear condition in attack of monsters?

nevermind

dudantas committed:
Merge branch 'main' into fear_condition

paczielny commented:
![image](https://github.com/opentibiabr/canary/assets/125104443/624ff5bf-0366-460e-a287-35d4c6da2dec)

compiled in 13.16

paczielny commented:
> 

![image](https://github.com/opentibiabr/canary/assets/125104443/ef6bea8f-02eb-4c87-a295-a3c82faa5d53)

dudantas committed:
fix: register condition feared

dudantas commented:
> ![image](https://user-images.githubusercontent.com/125104443/237536199-624ff5bf-0366-460e-a287-35d4c6da2dec.png)
> 
> compiled in 13.16

Please, test again with my last commit. https://github.com/opentibiabr/canary/pull/542/commits/11892d61e205534b15fd76ac7a51b5f8becf61d6

paczielny commented:
> > ![image](https://user-images.githubusercontent.com/125104443/237536199-624ff5bf-0366-460e-a287-35d4c6da2dec.png)
> > compiled in 13.16
> 
> Please, test again with my last commit. [11892d6](https://github.com/opentibiabr/canary/commit/11892d61e205534b15fd76ac7a51b5f8becf61d6)

working perfectly!  thanks @dudantas

dudantas committed:
fix: resolve sonar checks and improve some codes

dudantas committed:
improve: extract code in a separated function for better understanding and reduce complexity
